Refactor exec_jec

uncomputable opened the pull request:
A small change that fits into no other PR.

uncomputable commented:
Added another bugfix that fits nowhere else.

apoelstra commented on src/bit_machine/mod.rs:
```diff
@@ -365,67 +369,74 @@ impl BitMachine {
-        use simplicity_sys::c_jets::uword_width;
-        use simplicity_sys::ffi::UWORD;
+    /// Return a read frame that is equivalent to the active read frame.
+    fn get_input_frame(&mut self, bit_width: usize, buffer: &mut Vec<UWORD>) -> CFrameItem {
```
In 63c2d233fb29a8837e930d79507d89d06a4443d9:

`get_input_frame` and `get_output_frame` both need to be marked `unsafe` since they return a data structure containing raw pointers into a `Vec`, with no protection on those pointers being invalidated.

Probably these methods should just be defined inside of `exec_jet` since they're so wildly unsafe to use.

uncomputable replied:
I moved the new functions inside `exec_jet`. Because these functions are pure, we need to pass `self` as a parameter. I think this is a better solution than closures which cannot be unsafe.

apoelstra commented on src/bit_machine/mod.rs:
```diff
@@ -365,67 +369,74 @@ impl BitMachine {
+    fn get_input_frame(&mut self, bit_width: usize, buffer: &mut Vec<UWORD>) -> CFrameItem {
+        let buffer_width = uword_width(bit_width);
+        buffer.clear();
+        buffer.resize(buffer_width, 0);
```
In 63c2d233fb29a8837e930d79507d89d06a4443d9:

I would instead accept a `&mut [u8]` and `assert_eq` that its length is correct, and make the caller do the resizing and initialization. Then at least we're not invalidating other `CFrameItem`s that refer to this object from within this function.

uncomputable replied:
I will look into this tomorrow

apoelstra commented on src/bit_machine/mod.rs:
```diff
@@ -365,67 +369,74 @@ impl BitMachine {
+    }
+
+    /// Return the value of the given type that is contained in the buffer.
+    fn read_output_frame(&mut self, bit_width: usize, buffer: &mut Vec<UWORD>) {
```
In 63c2d233fb29a8837e930d79507d89d06a4443d9:

Here you can take a `&[UWORD]` and you don't even need `mut`.

apoelstra commented on src/bit_machine/mod.rs:
```diff
@@ -366,66 +367,79 @@ impl BitMachine {
+        use crate::ffi::c_jets::uword_width;
+        use crate::ffi::ffi::UWORD;
+
+        /// Return a read frame that is equivalent to the active read frame.
```
In 494a232fa8e6144be845940a1bfe22c88e98359e:

I'm much happier now that this lives inside `exec_jet` and is clearly a single-use function. But we should improve the comment still, to say something like
```
/// Creates a new read frame, backed by `buffer`, and copy the `bit_width` bits from the active read frame into it.
///
/// The value of `buffer`, including possibly its backing store, will be lost and no pointers
/// to it should be retained. Similarly the returned frame should not outlive `buffer` or
/// any modifications to it.
///
/// `bit_width` should be the width of the active read frame.
```

apoelstra commented on src/bit_machine/mod.rs:
```diff
@@ -366,66 +367,79 @@ impl BitMachine {
-        if a_frame_size == 0 && b_frame_size == 0 {
-            return Ok(());
+
+        /// Return a write frame that is equivalent to the active write frame.
```
In 494a232fa8e6144be845940a1bfe22c88e98359e:

This comment should just say "returns a new output frame backed by `buffer` of width `bit_width`". And have similar warnings about the safety of `buffer` and maybe say that `bit_width` is expected to be the size of the active write frame.

apoelstra commented on src/bit_machine/mod.rs:
```diff
@@ -366,66 +367,79 @@ impl BitMachine {
-                c_writeBit(&mut a_frame, bit);
+
+        /// Return the value of the given type that is contained in the buffer.
+        fn read_output_frame(mac: &mut BitMachine, bit_width: usize, buffer: &[UWORD]) {
```
In 494a232fa8e6144be845940a1bfe22c88e98359e:

Could this simply be a method on `Frame`? The frame itself should know its width now.

As written, is this unsafe if the user passes too large a value for `bit_width`?

uncomputable replied:
`Frame`s don't own their data so we would have to pass the data as a parameter, which is messy.

uncomputable commented:
Rebased on master. Creating the buffers inside the helper methods is much cleaner than passing mutable references. Added assertions that the active read / write frame are large enough. Resolved misunderstanding that active read / write frame is as wide as jet source / target type.

apoelstra commented on src/bit_machine/mod.rs:
```diff
@@ -366,66 +380,93 @@ impl BitMachine {
+        /// ## Panics
+        ///
+        /// Active read frame has fewer bits than `bit_width`.
+        unsafe fn get_input_frame(
```
In b561354638232b5884f8197a47e8cfda415a16f5:

Ok, awesome, this is looking much cleaner.

Should add a `## Safety` section to the doc simply saying that it returns a `CFrameItem` which points into the returned `Vec` and that the `Vec` must outlive the `CFrameItem`.

*Alternately* we could create a new opaque type `struct BackedCFrameItem` which holds the `Vec`, implements `Deref<CFrameItem>`, and which would then ensure that you were obeying the lifetime rules. I *think* that if you did this, you could eliminate `unsafe` from this method.

uncomputable replied:
Ok, will do. I haven't been working much with unsafe, so these comments on how to write unsafe Rust (somewhat) safely are appreciated :)

apoelstra commented on src/bit_machine/mod.rs:
```diff
@@ -366,66 +380,93 @@ impl BitMachine {
+        /// ## Panics
+        ///
+        /// Active write frame has fewer bits than `bit_width`.
+        fn update_active_write_frame(mac: &mut BitMachine, bit_width: usize, buffer: &[UWORD]) {
```
In b561354638232b5884f8197a47e8cfda415a16f5:

Here you still need `unsafe` because the passed buffer might now be wide enough for your `c_readBit` calls. I think you can add another `assert!` that `buffer.len()` is the right length to avoid this unsafety.

apoelstra reviewed (approved):
ACK 44d20a96f123fc48f7b3007b962a8e7f8cb7cc99
